aiida_vasp/parsers/vasp.py

```diff
@@ -17,78 +115,166 @@ class VaspParser(BaseParser):
 
     def __init__(self, calc):
         super(VaspParser, self).__init__(calc)
+
         self.out_folder = None
-        self.vrp = None
-        self.dcp = None
+        self._settings = DEFAULT_OPTIONS
+
+        try:
+            self._settings.update(self._calc.inp.settings.get_dict()['parser_settings'])
+        except KeyError:
+            # There are no special parser settings so we just return the default settings
+            pass
+
+        self._check_and_validate_settings()
+
+        self._nodes_to_add = list(PARSABLE_QUANTITIES.keys())
+        self._parsable_files = PARSABLE_FILES
+
+        self._parsers = {
+            'vasprun.xml': None,
+            'DOSCAR': None,
+            'IBZKPT': None,
+            'OUTCAR': None,
+            'EIGENVAL': None,
+        }
+
+        self._quantities_to_parse = []
+        self._output_nodes = {}
 
     def parse_with_retrieved(self, retrieved):
+
         self.check_state()
         self.out_folder = self.get_folder(retrieved)
+
         if not self.out_folder:
             return self.result(success=False)
-        outcar = self.get_file('OUTCAR')
-        if not outcar:
-            self.logger.error('OUTCAR not found, ' + 'look at the scheduler output for troubleshooting')
-            return self.result(success=False)
 
-        self.vrp = self.read_run()
+        # Get all specialised file parsers. Warnings will be issued if a file should be parsed and
+        # the corresponding files do not exist.
+        success = self._set_file_parsers()
 
-        bands, kpout, structure = self.read_eigenval()
+        if not success:
+            # A critical file i.e. OUTCAR does not exist. Abort parsing.
+            return self.result(success=False)
 
-        self.dcp = self.read_dos()
-        dosnode = self.get_dos_node(self.vrp, self.dcp)
+        # Get an initial list of quantities which should be parsed.
+        self._update_parsing_list()
 
-        if bands:
-            self.set_bands(bands)  # append output nodes
+        # Parse all implemented quantities in the nodesToAdd list, if they should be parsed. The list
+        # might get dynamically updated during the loop.
+        while self._quantities_to_parse:
+            quantity = self._quantities_to_parse.pop(0)
+            if self._settings['add_' + quantity]:
```

Comment:
Can you please add the usage of parser_settings into the docstring of the class?
Something like
```
"""
...
Parser Settings usage:

Parser settings can be passed through the input node `settings` as follows::

  settings = ParameterData(dict={
    'parser_settings': {
        ...
    }
  })

Valid keys for `parser_settings` are:

 * `add_<quantity>`, where quantity one of .... This causes ...
 * ...
"""
```

Reply:
I added the docketing with some additional information what the particular keys will do. I think a shorter version of that or at least a hint on where to look for it should also go to the VaspCalculation docstring, because 'settings' is an attribute of VaspCalculation and people might more likely read that one.